### ISSUE TITLE
Non-regression test harness

### ISSUE THREAD

**Issue by arey** [spring-petclinic/spring-petclinic-rest#140]
The Swagger UI provides a UI to do manual testing. 

In order to be able to make changes or refactoring with peace of mind, an automated test set with assertions could be useful.
I'm thinking of a Postman collection, for example. But any other idea is welcome.

**Comment by Isaac-Andradee** [comment]
Thats a great idea, @arey .

We can implement the non-regression test harness using a Postman collection with Newman integration. But i think that we need to ensure that we fully cover all necessary aspects, could you please provide some additional details? Like:

- **Critical Endpoints & Features:**
  - Which endpoints and functionalities should be prioritized in the tests or ignored?
  - Are there any interdependent endpoints that should be tested together? 
- **Authentication & Authorization:**
  - What authentication method is being used (e.g., JWT, API keys)? How should the tests handle the login/authentication flow?
- **Test Data & Scenarios:**
  - Are there any specific test data sets we should use? Should we include both positive and negative test cases (e.g., valid requests vs. invalid inputs)?
- **Acceptance Criteria & Assertions:**
  - What are the expected outcomes for each endpoint? Could you provide examples of expected HTTP statuses, response structures, or error messages?
- **CI/CD Integration:**
  - How do you envision integrating these tests into our current CI/CD pipeline? Are there any specific reporting formats or tools that we should follow?
- Are there any special considerations, performance constraints, or security guidelines we need to adhere to?

**PS**: I know that some of these questions can be answered by reading the code directly, but I leave them explicitly here to ensure that no critical part has gone unnoticed or to discover if there are special behaviors not evident just by reading the code. Anyway, to avoid misunderstandings.

Thank's.

**Comment by arey** [comment]
Hi @Isaac-Andradee, lots of good questions.

- We could start simpler with nominal use cases: those used by the Angular version of Spring Petclinic to add owners, pets and visits.
- With the exception of pet types, we don't have to rely on existing data. The postman scenario has to create new owner, visit and pet test cases before altering them.
- From my point of view, the assertions need to test the HTTP status and some well-chosen data (not all the form data, because that is already tested in the unit test).
- Later we could add a CI/CD Newman workflow. Maybe in manual mode?

**Comment by Isaac-Andradee** [comment]
Yeah. Sounds good your proposed approach. Let me show you what i'm thinking to do on this enhancement...

1.  Nominal Flow Testing
    - Creating an Owner: Implement a POST request to create a new owner.
    - Creating a Pet: Use the returned owner ID from the previous request to create a new pet (utilizing an already existing pet type).
    - Creating a Visit: With the newly created pet, perform a POST request to register a visit, ensuring that the key response data (e.g., IDs) is correctly validated.
2. Assertions and Test Validations
    - HTTP Status Checks: Each request will assert the proper HTTP status (e.g., 201 Created).
    - Key Data Assertions: Tests will also validate essential response properties (e.g., the existence of an owner or pet ID) using Postman test scripts, without over-testing the complete form data already covered by unit tests.
3. Project File and Test Organization
    - Postman Collection: Create a Postman collection structured with folders (e.g., "Owners", "Pets", "Visits") to logically group these tests.
    - Environment Setup: We can configure a Postman environment with variables such as {{baseUrl}}, {{ownerId}}, and {{petId}} to allow chaining between requests.
    - Newman Execution Script: We can develop a shell script (e.g., `run-tests.sh`) that enables manual test execution via Newman, such as:
``` bash 
newman run myCollection.json -e myEnvironment.json --reporters cli,html --timeout-request 5000
```
> note: this command above  is just a generic one.

What you think about it? 

PS: Probably we'll need to update the project doc to include instructions on executing tests manually with Newman?

**Comment by arey** [comment]
Thanks for your proposal @Isaac-Andradee. Sounds like a good idea to me.

In term of organisation, I suggest giving the test scenarios functional names. For example, the `Nominal Flow Test` you are referring to could be called "First visit" and have the description "A new customer comes wit her pet for a first visit", or something like that.
In the future, we could have a second scenario : "A customer returns for a new visit, updates their details and those of their pet".
This is a more BDD approach. But it could be debated and discussed. What do you think about?

**Comment by Isaac-Andradee** [comment]
Totally agreed, @arey . 

I'll go ahead with this line of reasoning and find some time to resolve this one calmly... 

You can assign me to this issue :)
